source/portable/NetworkInterface/DriverSAM/NetworkInterface.c

```diff
@@ -805,29 +819,94 @@ static uint16_t prvGenerateCRC16( const uint8_t * pucAddress )
     usSum ^= ( usValues[ 4 ] >> 4 ) ^ ( usValues[ 4 ] << 2 );
     usSum ^= ( usValues[ 5 ] >> 2 ) ^ ( usValues[ 5 ] << 4 );
 
-    usSum &= 0x3FU;
+    usSum &= GMAC_ADDRESS_HASH_MASK;
     return usSum;
 }
 /*-----------------------------------------------------------*/
 
-static void prvAddMulticastMACAddress( const uint8_t * ucMacAddress )
+/**
+ * @brief Adds a multicast mac address to the GMAC's hash register. Internally, this function
+ * keeps track of the hashes of MAC addresses being passed and it also keeps a counter for how many
+ * times every bit in the GMAC hash register has been hit. This way, if two multicast MAC
+ * addresses map to the same GMAC hash register bit, the internal counter will hold the value of 2.
+ * Calling prvRemoveMulticastMACAddress for one of those MACs will only decrement the counter, but
+ * the bit in the GMAC hash register will remain set until a call to prvRemoveMulticastMACAddress for
+ * the other MAC causes the counter to reach 0 and the bit to get cleared.
+ *
+ * @param[in] pucMacAddress: A pointer to the multicast MAC Address in question.
+ */
+static void prvAddMulticastMACAddress( const uint8_t * pucMacAddress )
 {
-    uint32_t ulMask;
-    uint16_t usIndex;
+    /* Note: Only called from the IPTask, so no thread-safety is required. */
+    uint8_t ucBinNumber;
 
-    usIndex = prvGenerateCRC16( ucMacAddress );
+    if( NULL == pucMacAddress )
+    {
+        return;
+    }
 
-    ulMask = 1U << ( usIndex % 32 );
+    FreeRTOS_debug_printf( "prvAddMulticastMACAddress: %02X-%02X-%02X-%02X-%02X-%02X",
+                           pucMacAddress[ 0 ], pucMacAddress[ 1 ], pucMacAddress[ 2 ], pucMacAddress[ 3 ], pucMacAddress[ 4 ], pucMacAddress[ 5 ] );
+    ucBinNumber = prvGenerateCRC16( pucMacAddress );
 
-    if( usIndex < 32U )
+    /* If the bin counter corresponding to this mac address is already non-zero, */
+    /* a multicast address with the same hash has already been added, so there's nothing more to do. */
+    if( 0 == prvAddressHashCounters[ ucBinNumber ] )
     {
-        /* 0 .. 31 */
-        GMAC->GMAC_HRB |= ulMask;
+        /* This bin counter is zero, so this is the first time we are registering a MAC with this hash. */
+        prvAddressHashBitMask |= ( uint64_t ) ( ( uint64_t ) 1 << ucBinNumber );
+        gmac_set_hash64( GMAC, prvAddressHashBitMask );
+        gmac_enable_multicast_hash( GMAC, true );
```

Comment:
Mind you that some Atmel implementations of `gmac_enable_multicast_hash()` are erroneous:
~~~
static inline void gmac_enable_multicast_hash(Gmac* p_gmac, uint8_t uc_enable)
{
	if (uc_enable) {
		p_gmac->GMAC_NCFGR |= GMAC_NCFGR_UNIHEN;
	} else {
		p_gmac->GMAC_NCFGR &= ~GMAC_NCFGR_UNIHEN;
	}
}
~~~
`GMAC_NCFGR_MTIHEN` should be used in stead of `GMAC_NCFGR_UNIHEN`.

The function `gmac_enable_unicast_hash()` was never created, I don't know why.

Shouldn't we enable/disable both multicast and unicast filtering?

Reply:
Thanks for reminding me about that one. I had a note about this and I do enable the proper multicast bit in `prvGMACInit()` but I somehow forgot to fix the calls you mentioned.

BTW, is there a good reason not to fix/add those functions in gmac_SAM.c? I know it is copied from the Advanced Software Framework, but we know these are errors for sure.... Should't I just fix them and move on with using proper functions like `gmac_enable_multicast_hash()` and `gmac_enable_unicast_hash()` instead of being forced to resort to setting the individual bits like `GMAC_NCFGR_UNIHEN`?